spynnaker/pyNN/models/neuron/synapse_dynamics/synapse_dynamics_structural.py

```diff
@@ -0,0 +1,514 @@
+        presynaptic_machine_vertices = []
+        structural_application_edges = []
+        structural_machine_edges = []
+        max_subpartitions = 0
+
+        for app_edge in application_graph.get_edges_ending_at_vertex(
+                app_vertex):
+            if isinstance(app_edge, ProjectionApplicationEdge):
+                for synapse_info in app_edge.synapse_information:
+                    if synapse_info.synapse_dynamics is self:
+                        structural_application_edges.append(app_edge)
+                        population_to_subpopulation_information[
+                            app_edge.pre_vertex] = []
+                        break
+
+        no_pre_populations = len(structural_application_edges)
+        # For each structurally plastic APPLICATION edge find the c
+        # orresponding machine edges
+        for machine_edge in machine_graph.get_edges_ending_at_vertex(
+                machine_vertex):
+            if isinstance(machine_edge, ProjectionMachineEdge):
+                for synapse_info in machine_edge._synapse_information:
+                    if synapse_info.synapse_dynamics is self:
+                        structural_machine_edges.append(machine_edge)
+                        # For each structurally plastic MACHINE edge find the
+                        # corresponding presynaptic subvertices
+                        presynaptic_machine_vertices.append(
+                            machine_edge.pre_vertex)
+
+        # For each presynaptic subvertex figure out the partition (?)
+        # to retrieve the key and n_atoms
+        for vertex in presynaptic_machine_vertices:
+            population_to_subpopulation_information[
+                graph_mapper.get_application_vertex(vertex)].append(
+                (routing_info.get_routing_info_from_pre_vertex(
+                    vertex, constants.SPIKE_PARTITION_ID).first_key,
+                 graph_mapper.get_slice(vertex)[2],
+                 graph_mapper.get_slice(vertex)[0],
+                 routing_info.get_routing_info_from_pre_vertex(
+                     vertex, constants.SPIKE_PARTITION_ID).first_mask))
+
+        for subpopulation_list in \
+                population_to_subpopulation_information.itervalues():
+            max_subpartitions = np.maximum(max_subpartitions,
+                                           len(subpopulation_list))
+
+        # Current machine vertex key (for future checks)
+        current_key = routing_info.get_routing_info_from_pre_vertex(
+            machine_vertex, constants.SPIKE_PARTITION_ID)
+
+        if current_key is not None:
+            current_key = current_key.first_key
+        else:
+            current_key = -1
+
+        # Table header
+        spec.write_value(data=no_pre_populations, data_type=DataType.INT32)
+
+        total_words_written = 0
+        for subpopulation_list in \
+                population_to_subpopulation_information.itervalues():
```

Comment:
This line is not Python-3 ready. We probably ought to make sure the code isn't making that sort of problem for us.
Use `itervalues(population_to_subpopulation_information)` (where `itervalues` comes from the `six` package).